src/components/loading_indicator.rs

```diff
@@ -0,0 +1,93 @@
+use gtk::glib;
+use gtk::prelude::*;
+use gtk::subclass::prelude::*;
+
+mod imp {
+    use super::*;
+    use gtk::graphene;
+    use std::cell::Cell;
+
+    #[derive(Default)]
+    pub struct LoadingIndicator {
+        pub(super) start_time: Cell<i64>,
+        pub(super) progress: Cell<f64>,
+    }
+
+    #[glib::object_subclass]
+    impl ObjectSubclass for LoadingIndicator {
+        const NAME: &'static str = "ComponentsLoadingIndicator";
+        type Type = super::LoadingIndicator;
+        type ParentType = gtk::Widget;
+
+        fn class_init(klass: &mut Self::Class) {
+            klass.set_css_name("loadingindicator");
+        }
+    }
+
+    impl ObjectImpl for LoadingIndicator {
+        fn constructed(&self) {
+            self.obj().connect_visible_notify(|widget| {
+                if widget.is_visible() {
+                    widget.imp().start_time.set(widget.time());
+                    widget.add_tick_callback(|widget, _clock| {
+                        widget.queue_draw();
+                        Continue(widget.is_visible())
+                    });
+                }
+            });
+        }
+    }
+
+    impl WidgetImpl for LoadingIndicator {
+        fn measure(&self, _orientation: gtk::Orientation, for_size: i32) -> (i32, i32, i32, i32) {
+            (0, for_size, -1, -1)
+        }
+
+        fn snapshot(&self, snapshot: &gtk::Snapshot) {
+            let widget = self.obj();
+            let size = widget.width() as f32;
+            let bounds = graphene::Rect::new(0.0, 0.0, size, size);
+
+            let context = snapshot.append_cairo(&bounds);
+            let color = widget.color();
+            context.set_source_rgba(
+                color.red() as _,
+                color.green() as _,
+                color.blue() as _,
+                color.alpha() as _,
+            );
+            let half_size = size as f64 / 2.0;
+
+            let pi = std::f64::consts::PI;
+
+            context.set_line_width(2.0);
+
+            let time = widget.time() - self.start_time.get();
+            let shift = (time as f64 / 300000.0) % (2.0 * pi);
+
+            let start = shift - 0.5 * pi;
+            let diff = self.progress.get().max(0.01) * 2.0 * pi;
+
+            context.arc(half_size, half_size, half_size - 4.0, start, start + diff);
```

Comment:
I guess the `- 4.0` is for the margin? I prefer having that defines in CSS, and defined by the "users" of this widget (so, in the messagedocument part in this case).

Reply:
![image](https://user-images.githubusercontent.com/63008755/227613325-188cc6aa-07f3-40d7-8fb4-1301332bc2db.png)

Reply:
2.0 is enough

![image](https://user-images.githubusercontent.com/63008755/227613533-ba8e9265-c894-41e0-9af8-4510abe190f2.png)